@bknight1 — CI on this PR is failing, two things to address:

### 1. Rebase onto current `development` (picks up a race fix you don't have)

This PR's base is `848b8b9` (the #178 merge commit). Yesterday after that landed, three more PRs were merged:
- #177 — `Swarm.advection` `max_its=10` fix (your #182 also has this; will conflict).
- #179 — `memprobe` diagnostic module + KDTree counters.
- A 1-line race fix `5f64de2` for `UWexpression._ephemeral_expr_names` iteration mid-finalize (the "dictionary changed size during iteration" race).

Your CI shows the race fix matters here:
```
FAILED tests/test_0810_bc_unit_conversion.py::test_bc_accepts_raw_numbers
  - RuntimeError: dictionary changed size during iteration
```

Rebase:
```bash
git fetch origin
git rebase origin/development
# resolve any conflicts (likely small — your changes are in different files)
git push --force-with-lease
```

### 2. Investigate the integral / swarm-statistics regression

CI also reports:
```
FAILED tests/test_0501_integrals.py::test_integrate_meshvar           - assert 0.6366197... < 0.001
FAILED tests/test_0501_integrals.py::test_integrate_swarmvar_deriv_O1 - assert 2.0 < 0.001
FAILED tests/test_0501_integrals.py::test_integrate_swarmvar_deriv_03 - assert 2.0 < 0.001
FAILED tests/test_0502_boundary_integrals.py::test_bd_integral_meshvar - Expected 0.6366..., got 0.0
FAILED tests/test_0852_swarm_integration_statistics.py::...           - assert 0.0 > 0
FAILED tests/test_0852_swarm_integration_statistics.py::...           - assert 273.0 < 0.0
```

These look like *real regressions caused by this PR*, not pre-existing flakes:
- `test_bd_integral_meshvar` returning `0.0` instead of `0.6366...` is severe — a boundary integral computing zero is the kind of thing you only see if data isn't reaching the integration context at all.
- `test_integrate_meshvar` returning `0.6366` where the test expects `< 0.001` (an error-norm assertion) is the same shape — the *raw value* leaking through where the *residual* should be.

My guess: the `array.owner` existence guard in your `MeshVariable` callbacks is silently early-returning when it shouldn't (e.g. during a teardown-and-rebuild of an integration work-variable, where `owner` is briefly absent but the data sync is still required). Or the `mesh_update_lock` is gating a path that integration relies on.

Worth running these four tests locally with `UW_MEMPROBE=1` and a print/breakpoint in the new callback guard to see which path is being short-circuited.

After the rebase + fix, push and CI should pass — at which point we can merge.

--- a/src/underworld3/cython/petsc_types.pyx
+++ b/src/underworld3/cython/petsc_types.pyx
@@ -1,9 +1,36 @@
-from libc.stdlib cimport malloc
+from libc.stdlib cimport malloc, free
 
 cdef class PtrContainer:
 
+    def __cinit__(self):
+        self.fns_residual = NULL
+        self.fns_bcs = NULL
+        self.fns_jacobian = NULL
+        self.fns_bd_residual = NULL
+        self.fns_bd_jacobian = NULL
+
+    def __dealloc__(self):
+        if self.fns_residual != NULL:
+            free(self.fns_residual)
+        if self.fns_bcs != NULL:
+            free(self.fns_bcs)
+        if self.fns_jacobian != NULL:
+            free(self.fns_jacobian)
+        if self.fns_bd_residual != NULL:
+            free(self.fns_bd_residual)
+        if self.fns_bd_jacobian != NULL:
+            free(self.fns_bd_jacobian)
+
     cpdef allocate(self, int n_res, int n_bcs, int n_jac, int n_bd_res, int n_bd_jac):
         """Allocate function pointer arrays of the given sizes."""
+
+        # Free existing memory if already allocated
+        if self.fns_residual != NULL: free(self.fns_residual)
+        if self.fns_bcs != NULL:      free(self.fns_bcs)
+        if self.fns_jacobian != NULL: free(self.fns_jacobian)
+        if self.fns_bd_residual != NULL: free(self.fns_bd_residual)
+        if self.fns_bd_jacobian != NULL: free(self.fns_bd_jacobian)
+
         self.fns_residual    = <PetscDSResidualFn*>   malloc(n_res    * sizeof(PetscDSResidualFn))
         self.fns_bcs         = <PetscDSResidualFn*>   malloc(n_bcs    * sizeof(PetscDSResidualFn))
         self.fns_jacobian    = <PetscDSJacobianFn*>   malloc(n_jac    * sizeof(PetscDSJacobianFn))

--- a/src/underworld3/discretisation/discretisation_mesh.py
+++ b/src/underworld3/discretisation/discretisation_mesh.py
@@ -594,14 +594,18 @@ class Mesh(Stateful, uw_object):
         # to handle that so we just wrap it here.
 
         def mesh_update_callback(array, change_context):
+            mesh = array.owner
+            if mesh is None:
+                return
+
             print(f"Mesh update callback - mesh deform")
-            coords = array.reshape(-1, array.owner.cdim)
-            self._deform_mesh(coords, verbose=True)
+            coords = array.reshape(-1, mesh.cdim)
+            mesh._deform_mesh(coords, verbose=True)
 
             # Increment mesh version to notify registered swarms of coordinate changes
-            with self._mesh_update_lock:
-                self._mesh_version += 1
-                print(f"Mesh version incremented to {self._mesh_version}")
+            with mesh._mesh_update_lock:
+                mesh._mesh_version += 1
+                print(f"Mesh version incremented to {mesh._mesh_version}")
 
             return
 
@@ -1337,10 +1341,14 @@ class Mesh(Stateful, uw_object):
             )
 
             def mesh_update_callback(array, change_context):
-                coords = array.reshape(-1, array.owner.cdim)
-                self._deform_mesh(coords, verbose=False)
-                with self._mesh_update_lock:
-                    self._mesh_version += 1
+                mesh = array.owner
+                if mesh is None:
+                    return
+
+                coords = array.reshape(-1, mesh.cdim)
+                mesh._deform_mesh(coords, verbose=False)
+                with mesh._mesh_update_lock:
+                    mesh._mesh_version += 1
                 return
 
             self._coords.add_callback(mesh_update_callback)

--- a/src/underworld3/discretisation/discretisation_mesh.py
+++ b/src/underworld3/discretisation/discretisation_mesh.py
@@ -285,6 +285,11 @@ class Mesh(Stateful, uw_object):
         self._registered_swarms = weakref.WeakSet()
         self._registered_surfaces = weakref.WeakSet()  # Surfaces using this mesh
         self._registered_submeshes = weakref.WeakSet()  # Submeshes from extract_region
+
+        # _mesh_update_lock: Re-entrant lock to coordinate mesh deformation.
+        # Held by mesh_update_callback during _deform_mesh(). Checked by
+        # MeshVariable callbacks (blocking=False) to skip PETSc sync during
+        # sensitive coordinate changes.
         self._mesh_update_lock = threading.RLock()
 
         comm = PETSc.COMM_WORLD
@@ -1785,42 +1790,43 @@ class Mesh(Stateful, uw_object):
         The coord array that is passed in should match the shape of self.data
         """
 
-        coord_vec = self.dm.getCoordinatesLocal()
-        coords = coord_vec.array.reshape(-1, self.cdim)
-        coords[...] = new_coords[...]
+        with self._mesh_update_lock:
+            coord_vec = self.dm.getCoordinatesLocal()
+            coords = coord_vec.array.reshape(-1, self.cdim)
+            coords[...] = new_coords[...]
 
-        self.dm.setCoordinatesLocal(coord_vec)
-        self.nuke_coords_and_rebuild()
+            self.dm.setCoordinatesLocal(coord_vec)
+            self.nuke_coords_and_rebuild()
 
-        # Rebuild the _coords array view.  nuke_coords_and_rebuild may
-        # replace the coordinate vector internally (createCoordinateSpace),
-        # leaving self._coords as a stale numpy view of the old buffer.
-        import underworld3.utilities
-        old_callbacks = getattr(self._coords, "_callbacks", [])
-        self._coords = underworld3.utilities.NDArray_With_Callback(
-            numpy.ndarray.view(
-                self.dm.getCoordinatesLocal().array.reshape(-1, self.cdim)
-            ),
-            owner=self,
-        )
-        for cb in old_callbacks:
-            self._coords.add_callback(cb)
+            # Rebuild the _coords array view.  nuke_coords_and_rebuild may
+            # replace the coordinate vector internally (createCoordinateSpace),
+            # leaving self._coords as a stale numpy view of the old buffer.
+            import underworld3.utilities
+            old_callbacks = getattr(self._coords, "_callbacks", [])
+            self._coords = underworld3.utilities.NDArray_With_Callback(
+                numpy.ndarray.view(
+                    self.dm.getCoordinatesLocal().array.reshape(-1, self.cdim)
+                ),
+                owner=self,
+            )
+            for cb in old_callbacks:
+                self._coords.add_callback(cb)
 
-        # BUGFIX(#122): mark registered solvers for rebuild. Since PR #127
-        # ("Trust JIT cache: skip DM rebuild on constant-only parameter
-        # changes") a solver with is_setup=True trusts its cached PETSc DM
-        # / SNES assembly and skips rebuild on the next solve(). After a
-        # coordinate change the cached DM still carries pre-deform
-        # coordinates, so F(v_prev) ≈ 0 and the solver converges in 0
-        # iterations without updating the solution. mesh.adapt() already
-        # does this; _deform_mesh must match.
-        for solver in self._equation_systems_register:
-            if solver is not None and hasattr(solver, "is_setup"):
-                solver.is_setup = False
+            # BUGFIX(#122): mark registered solvers for rebuild. Since PR #127
+            # ("Trust JIT cache: skip DM rebuild on constant-only parameter
+            # changes") a solver with is_setup=True trusts its cached PETSc DM
+            # / SNES assembly and skips rebuild on the next solve(). After a
+            # coordinate change the cached DM still carries pre-deform
+            # coordinates, so F(v_prev) ≈ 0 and the solver converges in 0
+            # iterations without updating the solution. mesh.adapt() already
+            # does this; _deform_mesh must match.
+            for solver in self._equation_systems_register:
+                if solver is not None and hasattr(solver, "is_setup"):
+                    solver.is_setup = False
 
-        # Propagate coordinate changes to registered submeshes
-        for submesh in self._registered_submeshes:
-            submesh.sync_coordinates_from_parent()
+            # Propagate coordinate changes to registered submeshes
+            for submesh in self._registered_submeshes:
+                submesh.sync_coordinates_from_parent()
 
         return
 

--- a/src/underworld3/discretisation/discretisation_mesh_variables.py
+++ b/src/underworld3/discretisation/discretisation_mesh_variables.py
@@ -494,6 +494,7 @@ class _BaseMeshVariable(Stateful, uw_object):
             """Callback to sync variable changes back to PETSc (like mesh.points)"""
             var = array.owner
             if var is None:
+                # Array is being accessed during object teardown (owner is gone)
                 return
 
             # Only act on data-changing operations (following mesh.points pattern)

--- a/src/underworld3/discretisation/discretisation_mesh_variables.py
+++ b/src/underworld3/discretisation/discretisation_mesh_variables.py
@@ -492,36 +492,40 @@ class _BaseMeshVariable(Stateful, uw_object):
         # Single callback function (following mesh_update_callback pattern)
         def variable_update_callback(array, change_context):
             """Callback to sync variable changes back to PETSc (like mesh.points)"""
+            var = array.owner
+            if var is None:
+                return
+
             # Only act on data-changing operations (following mesh.points pattern)
             data_changed = change_context.get("data_has_changed", True)
             if not data_changed:
                 return
 
             # Prevent recursion by checking if we're already in a callback
-            if hasattr(self, "_in_callback") and self._in_callback:
+            if hasattr(var, "_in_callback") and var._in_callback:
                 return
 
             # Set recursion guard
-            self._in_callback = True
+            var._in_callback = True
 
             try:
                 # Skip updates during mesh coordinate changes to prevent corruption
                 # Check if mesh is currently being updated
-                if hasattr(self.mesh, "_mesh_update_lock"):
+                if hasattr(var.mesh, "_mesh_update_lock"):
                     # Try to acquire lock without blocking - if we can't, skip update
-                    if not self.mesh._mesh_update_lock.acquire(blocking=False):
+                    if not var.mesh._mesh_update_lock.acquire(blocking=False):
                         return
                     try:
                         # Persist changes to PETSc (like mesh callback updates coordinates)
-                        self.pack_uw_data_to_petsc(array, sync=True)
+                        var.pack_uw_data_to_petsc(array, sync=True)
                     finally:
-                        self.mesh._mesh_update_lock.release()
+                        var.mesh._mesh_update_lock.release()
                 else:
                     # Fallback if no lock exists
-                    self.pack_uw_data_to_petsc(array, sync=True)
+                    var.pack_uw_data_to_petsc(array, sync=True)
             finally:
                 # Clear recursion guard
-                self._in_callback = False
+                var._in_callback = False
 
         # Register the callback (following mesh.points pattern)
         array_obj.add_callback(variable_update_callback)
@@ -556,34 +560,38 @@ class _BaseMeshVariable(Stateful, uw_object):
         # Callback for flat data format
         def flat_data_update_callback(array, change_context):
             """Callback to sync flat data changes back to PETSc"""
+            var = array.owner
+            if var is None:
+                return
+
             # Only act on data-changing operations
             data_changed = change_context.get("data_has_changed", True)
             if not data_changed:
                 return
 
             # Prevent recursion by checking if we're already in a callback
-            if hasattr(self, "_in_flat_callback") and self._in_flat_callback:
+            if hasattr(var, "_in_flat_callback") and var._in_flat_callback:
                 return
 
             # Set recursion guard
-            self._in_flat_callback = True
+            var._in_flat_callback = True
 
             try:
                 # Skip updates during mesh coordinate changes to prevent corruption
-                if hasattr(self.mesh, "_mesh_update_lock"):
-                    if not self.mesh._mesh_update_lock.acquire(blocking=False):
+                if hasattr(var.mesh, "_mesh_update_lock"):
+                    if not var.mesh._mesh_update_lock.acquire(blocking=False):
                         return
                     try:
                         # Use pack_raw for flat data format
-                        self.pack_raw_data_to_petsc(array, sync=True)
+                        var.pack_raw_data_to_petsc(array, sync=True)
                     finally:
-                        self.mesh._mesh_update_lock.release()
+                        var.mesh._mesh_update_lock.release()
                 else:
                     # Fallback if no lock exists
-                    self.pack_raw_data_to_petsc(array, sync=True)
+                    var.pack_raw_data_to_petsc(array, sync=True)
             finally:
                 # Clear recursion guard
-                self._in_flat_callback = False
+                var._in_flat_callback = False
 
         # Register the callback
         array_obj.add_callback(flat_data_update_callback)
@@ -2541,6 +2549,10 @@ class _BaseMeshVariable(Stateful, uw_object):
         # Single canonical callback for PETSc synchronization
         def canonical_data_callback(array, change_context):
             """ONLY callback that handles PETSc synchronization - prevents conflicts"""
+            var = array.owner
+            if var is None:
+                return
+
             # Only act on data-changing operations
             data_changed = change_context.get("data_has_changed", True)
             if not data_changed:
@@ -2556,26 +2568,26 @@ class _BaseMeshVariable(Stateful, uw_object):
 
             canonical_array = np.atleast_2d(array)
 
-            if canonical_array.shape != (canonical_array.shape[0], self.num_components):
+            if canonical_array.shape != (canonical_array.shape[0], var.num_components):
                 # Only reshape if we actually need to
-                canonical_array = canonical_array.reshape(-1, self.num_components)
+                canonical_array = canonical_array.reshape(-1, var.num_components)
 
             # Skip updates during mesh coordinate changes to prevent corruption
-            if hasattr(self.mesh, "_mesh_update_lock"):
-                if not self.mesh._mesh_update_lock.acquire(blocking=False):
+            if hasattr(var.mesh, "_mesh_update_lock"):
+                if not var.mesh._mesh_update_lock.acquire(blocking=False):
                     return
                 try:
                     # STEP 1: Sync to PETSc using established method with correct shape
-                    self.pack_raw_data_to_petsc(canonical_array, sync=True)
+                    var.pack_raw_data_to_petsc(canonical_array, sync=True)
                 finally:
-                    self.mesh._mesh_update_lock.release()
+                    var.mesh._mesh_update_lock.release()
             else:
                 # Fallback if no lock exists
-                self.pack_raw_data_to_petsc(canonical_array, sync=True)
+                var.pack_raw_data_to_petsc(canonical_array, sync=True)
 
             # STEP 2: Handle variable-specific updates (extensible like SwarmVariable)
-            if hasattr(self, "_on_data_changed"):
-                self._on_data_changed()
+            if hasattr(var, "_on_data_changed"):
+                var._on_data_changed()
 
         array_obj.add_callback(canonical_data_callback)
         return array_obj

--- a/src/underworld3/swarm.py
+++ b/src/underworld3/swarm.py
@@ -395,6 +395,9 @@ class SwarmVariable(DimensionalityMixin, MathematicalMixin, Stateful, uw_object)
             """Callback to sync variable changes back to PETSc (like swarm.points)"""
             var = array.owner
             if var is None:
+                # This guard handles cases where the array is modified during
+                # object teardown (e.g. at application exit), where the owning
+                # Python variable has already been garbage collected.
                 return
 
             # Only act on data-changing operations (following swarm.points pattern)

--- a/src/underworld3/swarm.py
+++ b/src/underworld3/swarm.py
@@ -393,17 +393,21 @@ class SwarmVariable(DimensionalityMixin, MathematicalMixin, Stateful, uw_object)
         # Single callback function (following swarm_update_callback pattern)
         def variable_update_callback(array, change_context):
             """Callback to sync variable changes back to PETSc (like swarm.points)"""
+            var = array.owner
+            if var is None:
+                return
+
             # Only act on data-changing operations (following swarm.points pattern)
             data_changed = change_context.get("data_has_changed", True)
             if not data_changed:
                 return
 
             # Skip updates during coordinate changes to prevent corruption
-            if hasattr(self.swarm, "_migration_disabled") and self.swarm._migration_disabled:
+            if hasattr(var.swarm, "_migration_disabled") and var.swarm._migration_disabled:
                 return
 
             # Persist changes to PETSc (like swarm callback updates coordinates)
-            self.pack_uw_data_to_petsc(array, sync=True)
+            var.pack_uw_data_to_petsc(array, sync=True)
 
         # Register the callback (following swarm.points pattern)
         array_obj.add_callback(variable_update_callback)

